frontend/packages/gitops-plugin/src/components/list/GitOpsTableRow.tsx

```diff
@@ -3,32 +3,110 @@ import i18n from '@console/internal/i18n';
 import * as classNames from 'classnames';
 import { GitOpsAppGroupData } from '../utils/gitops-types';
 import { RowFunction, TableData, TableRow } from '@console/internal/components/factory';
-import { Button } from '@patternfly/react-core';
-import { history, ExternalLink } from '@console/internal/components/utils';
+import { Button, Flex, FlexItem, Split, SplitItem, Tooltip } from '@patternfly/react-core';
+import { history, ExternalLink, Timestamp } from '@console/internal/components/utils';
 import { routeDecoratorIcon } from '@console/dev-console/src/components/import/render-utils';
+import {
+  GreenCheckCircleIcon,
+  YellowExclamationTriangleIcon,
+  GrayUnknownIcon,
+} from '@console/shared';
+import './GitOpsTableRow.scss';
 
 const tableColumnClasses = [
   classNames('pf-m-width-20'), // Application name
-  classNames('pf-m-width-40'), // Git repository
+  classNames('pf-m-width-30'), // Git repository
   classNames('pf-m-hidden', 'pf-m-visible-on-md', 'pf-m-width-20'), // Environments
-  classNames('pf-m-hidden', 'pf-m-visible-on-lg', 'pf-m-width-20'), // Last deployment
+  classNames('pf-m-hidden', 'pf-m-visible-on-lg', 'pf-m-width-30'), // Last deployment
 ];
 
-const handleClick = (appGroup: GitOpsAppGroupData) => {
+const handleClick = (appGroup) => {
   history.push(`/environments/${appGroup.name}?url=${appGroup.repo_url}`);
 };
 
+interface SyncProps {
+  tooltip: any[];
+  count: number;
+  icon: string;
+}
+
 const GitOpsTableRow: RowFunction<GitOpsAppGroupData> = (props) => {
   const { obj: appGroup, index, key, style } = props;
   const t = (tKey) => i18n.t(tKey);
+  let numOfSynced = 0;
+  let numOfOutOfSync = 0;
+  let numOfUnknown = 0;
+  const ttSynced = [];
+  const ttOutOfSync = [];
+  const ttUnknown = [];
+  let latestDeployedTime = '';
+  let latestDeployedEnv = '';
+  if (appGroup && appGroup.sync_status && appGroup.last_deployed) {
```

Comment:
I think you want to do this differently... being more data-driven is more functional. This feels bulky. I'll do up a changeset and see if I can clean this up some.

Reply:
For posterity, I chopped this up and treated it as data-first... the rendering happens later in the return statement. I've provided the code to Keith in Slack DMs... but I figured I should add it here as well so others can learn from it too. 

<details>
  <summary>The Replacement Component Rendering</summary>
  
  The diff is hard to read, so here is the final state:
```
const getMatchingEnvs = (envs: string[], desiredStatus: string) => (
  acc: string[],
  status: string,
  idx: number,
): string[] =>
  desiredStatus === status
    ? [...acc, envs[idx]] // 1:1 between a status and an env
    : acc;

const GitOpsTableRow: RowFunction<GitOpsAppGroupData> = (props) => {
  const { obj: appGroup, index, key, style } = props;
  const {
    name,
    sync_status: syncStatuses = [],
    environments: envs,
    last_deployed: lastDeployed = [],
    repo_url: repoUrl,
  } = appGroup;
  const t = (tKey) => i18n.t(tKey);

  const syncedEnvs: string[] = syncStatuses.reduce(getMatchingEnvs(envs, 'Synced'), []);
  const outOfSyncEnvs: string[] = syncStatuses.reduce(getMatchingEnvs(envs, 'OutOfSync'), []);
  const unknownEnvs: string[] = syncStatuses.reduce(getMatchingEnvs(envs, 'Unknown'), []);

  const latestDeployedTime = lastDeployed.reduce(
    (leadingDeployedTime, deployedTime) =>
      latestDeployedTime < deployedTime ? deployedTime : leadingDeployedTime,
    '',
  );
  const latestDeployedEnv = latestDeployedTime
    ? envs[lastDeployed.indexOf(latestDeployedTime)]
    : '';

  return (
    <TableRow id={index} index={index} trKey={key} style={style}>
      <TableData className={tableColumnClasses[0]}>
        <Button
          onClick={() => handleClick(appGroup)}
          aria-label={name}
          variant="link"
          style={{ padding: '0px' }}
        >
          {name}
        </Button>
      </TableData>
      <TableData className={classNames(tableColumnClasses[1])}>
        <ExternalLink href={repoUrl} additionalClassName={'co-break-all'}>
          <span style={{ marginRight: 'var(--pf-global--spacer--xs)' }}>
            {routeDecoratorIcon(repoUrl, 12, t)}
          </span>
          <span style={{ marginRight: 'var(--pf-global--spacer--xs)' }}>{repoUrl}</span>
        </ExternalLink>
      </TableData>
      <TableData className={classNames(tableColumnClasses[2], 'pf-u-text-nowrap')}>
        {syncStatuses ? (
          <Flex className="syncStatus">
            <GetSyncFragment
              tooltip={syncedEnvs.map((env) => (
                <Split hasGutter key={`${name}-${env}`}>
                  <SplitItem>
                    <GreenCheckCircleIcon />
                  </SplitItem>
                  <SplitItem isFilled style={{ color: 'white' }}>
                    {env}
                  </SplitItem>
                  <SplitItem style={{ color: 'white' }}>{t('gitops-plugin~Synced')}</SplitItem>
                </Split>
              ))}
              count={syncedEnvs.length}
              icon="check"
            />
            <GetSyncFragment
              tooltip={outOfSyncEnvs.map((env) => (
                <Split hasGutter key={`${name}-${env}`}>
                  <SplitItem>
                    <YellowExclamationTriangleIcon />
                  </SplitItem>
                  <SplitItem isFilled style={{ color: 'white' }}>
                    {env}
                  </SplitItem>
                  <SplitItem style={{ color: 'white' }}>{t('gitops-plugin~OutOfSync')}</SplitItem>
                </Split>
              ))}
              count={outOfSyncEnvs.length}
              icon="exclamation"
            />
            <GetSyncFragment
              tooltip={unknownEnvs.map((env) => (
                <Split hasGutter key={`${name}-${env}`}>
                  <SplitItem>
                    <GrayUnknownIcon />
                  </SplitItem>
                  <SplitItem isFilled style={{ color: 'white' }}>
                    {env}
                  </SplitItem>
                  <SplitItem style={{ color: 'white' }}>{t('gitops-plugin~Unknown')}</SplitItem>
                </Split>
              ))}
              count={unknownEnvs.length}
              icon="unknown"
            />
          </Flex>
        ) : (
          <span>{envs.join(', ')}</span>
        )}
      </TableData>
      <TableData className={tableColumnClasses[3]}>
        {latestDeployedTime ? (
          <Flex>
            <FlexItem className="lastDeploymentTime" spacer={{ default: 'spacerXs' }}>
              <span>
                <Timestamp timestamp={latestDeployedTime} />
              </span>
            </FlexItem>
            <FlexItem>{latestDeployedEnv}</FlexItem>
          </Flex>
        ) : (
          <span>-</span>
        )}
      </TableData>
    </TableRow>
  );
};
```
</details>

Reply:
I had to make one slight tweak.  Had to check that `syncStatus.length > 0` since it is an empty array initially, whereas before the sync status could be null.

Reply:
Thanks Andrew for the suggestion. And I thought that pushing the elements was elegant.